pennylane_lightning/src/rework/Gates.hpp

```diff
@@ -0,0 +1,298 @@
+// Copyright 2021 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file
+ * Defines quantum gates and their actions.
+ */
+#pragma once
+
+#include <vector>
+
+#include "typedefs.hpp"
+
+namespace Pennylane {
+
+    const double SQRT2INV = 0.7071067811865475;
+    const CplxType IMAG = CplxType(0, 1);
+
+    class AbstractGate {
+    protected:
+        const int numQubits;
+        const size_t length;
+        AbstractGate(int numQubits);
+    public:
+        /**
+         * @return the matrix representation for the gate as a one-dimensional vector.
+         */
+        virtual const std::vector<CplxType>& asMatrix() = 0;
+    };
+
+    // Single-qubit gates:
+
+    class SingleQubitGate : public AbstractGate {
+    protected:
+        SingleQubitGate();
+    };
+
+    class XGate : public SingleQubitGate {
+    private:
+        static const std::vector<CplxType> matrix;
+    public:
+        static const std::string label;
+        static XGate create(const std::vector<double>& parameters);
```

Comment:
Should the `create` method be kept, it would be worth providing a comment on what its purpose is.